lib/node/asar-fs-wrapper.ts

```diff
@@ -741,6 +741,53 @@ export const wrapFsWithAsar = (fs: Record<string, any>) => {
     return files;
   };
 
+  const { legacyMainResolve } = internalBinding('fs');
+  internalBinding('fs').legacyMainResolve = (packageJSONUrl: URL, packageConfigMain: string, base: string) => {
+    const pathInfo = splitPath(fileURLToPath(packageJSONUrl));
+    if (!pathInfo.isAsar) return legacyMainResolve(packageJSONUrl, packageConfigMain, base);
+    const { asarPath, filePath } = pathInfo;
+
+    const archive = getOrCreateArchive(asarPath);
+    if (!archive) {
+      throw createError(AsarError.INVALID_ARCHIVE, { asarPath });
+    }
+
+    // Copy out the original package.json into a temp directory so
+    // it can be read.
+    const tempPackageJSON = archive.copyFileOut(filePath);
```

Comment:
Was chatting with keeley about this, we shouldn't be copying the file out to a temp folder just to read it.  We should do what we do below and read from the asar fs at the offset

Reply:
> We should do what we do below and read from the asar fs at the offset

say a bit more? the impl here is picky about files existing at certain paths so i'm not sure we can drop in that approach quite the same 🤔

Reply:
Copying out from DMs:

we should patch the native method used to determine if a file exists / read the package json to use asar::Archive
* Extract the Archive class to a separate source_set
* Make lib_electron depend on that source_set
* Make electron_node depend on that source_set
* Patch the native method to support ASAR

like, copying a file out of the ASAR for every single module load is kinda crazy and gonna be way too slow

We need to just make node itself aware of ASAR instead of patching in a slow drifting JS impl